File: composeApp/src/androidMain/kotlin/beer/asmz/portfolio/kmp/ui/common/post/PostItem.kt
package beer.asmz.portfolio.kmp.ui.common.post

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Divider
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import beer.asmz.portfolio.kmp.R
import beer.asmz.portfolio.kmp.constant.BlurGrayColor
import beer.asmz.portfolio.kmp.constant.LightGrayColor
import beer.asmz.portfolio.kmp.constant.SubTextColor
import beer.asmz.portfolio.kmp.constant.TextColor
import beer.asmz.portfolio.kmp.model.Post
import beer.asmz.portfolio.kmp.model.PostContent
import beer.asmz.portfolio.kmp.model.PostContentPoster
import beer.asmz.portfolio.kmp.ui.AppTheme
import beer.asmz.portfolio.kmp.ui.common.chrome_tabs.launchCustomTabs
import coil3.compose.AsyncImage
import java.net.URL
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter


@Composable
fun PostItem(post: Post, onPressItem: ((post: Post) -> Unit)? = null) {
    val context = LocalContext.current

    val content = remember { post.content[0] }
    val poster = remember { content.poster?.get(0) }
    val hostname = remember { URL(content.url).host }
    val formattedDate = remember {
        val instant = Instant.ofEpochSecond(post.timestamp.toLong())
        val localDateTime = LocalDateTime.ofInstant(instant, ZoneId.systemDefault())
        val dtf = DateTimeFormatter.ofPattern("yyyy/MM/dd")
        dtf.format(localDateTime)
    }

    Column(
        modifier = Modifier
            .padding(vertical = 8.dp)
            .background(BlurGrayColor, shape = RoundedCornerShape(15.dp))
            .clip(RoundedCornerShape(15.dp))
            .fillMaxWidth()
            .clickable(onClick = {
                onPressItem?.let { onPress ->
                    onPress(post)
                } ?: run {
                    context.launchCustomTabs(content.url)
                }
            })
            .padding(horizontal = 16.dp, vertical = 32.dp)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            if (poster != null) {
                AsyncImage(
                    model = poster.url,
                    placeholder = painterResource(R.drawable.no_image),
                    contentDescription = content.title,
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .size(width = 112.dp, height = 88.dp)
                        .padding(end = 16.dp)
                )
            } else {
                Image(
                    painter = painterResource(id = R.drawable.no_image),
                    contentDescription = "no image",
                    contentScale = ContentScale.Fit,
                    modifier = Modifier
                        .size(width = 112.dp, height = 88.dp)
                        .padding(end = 16.dp)
                )
            }
            Column {
                Text(
                    text = content.title ?: "",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = TextColor,
                    modifier = Modifier.padding(bottom = 16.dp)
                )
                Text(hostname, color = SubTextColor)
            }
        }
        Row(
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            post.tags.forEach { tag ->
                Text(
                    text = "#$tag",
                    color = SubTextColor,
                    modifier = Modifier
                        .background(LightGrayColor, shape = RoundedCornerShape(5.dp))
                        .padding(vertical = 4.dp, horizontal = 8.dp)
                )
            }
        }
        Divider(modifier = Modifier.padding(vertical = 16.dp))
        Text(formattedDate, color = SubTextColor)
    }
}


@Preview
@Composable
fun PostItemPreview() {
    val post = Post(
        idString = "111",
        timestamp = 1672531200,
        tags = listOf("blog", "contribution"),
        content = listOf(
            PostContent(
                type = "link",
                description = "test description",
                title = "test title abc def ghi jkl mno pqr stu vwx yz - 123 456 789 0",
                url = "https://asmz.beer",
                poster = listOf(
                    PostContentPoster(url = "https://static.tumblr.com/nibuxm5/Q0yo2heyn/beer.jpg")
                )
            )
        )
    )

    AppTheme {
        PostItem(post)
    }
}

File: iosApp/iosApp/UI/Common/Post/PostItem.swift
//
//  PostItem.swift
//  iosApp
//
//  Created by 清水亮 on 2025/02/24.
//  Copyright © 2025 orgName. All rights reserved.
//

import Shared
import SwiftUI

struct PostItem: View {
    let post: Post
    let onPressItem: ((_ post: Post) -> Void)?

    var content: PostContent {
        post.content[0]
    }

    var poster: PostContentPoster? {
        content.poster?[0]
    }

    var hostname: String {
        if let url = content.url {
            return URL(string: url)?.host ?? ""
        } else {
            return ""
        }
    }

    var formattedDate: String {
        let unixtime = Double(post.timestamp)
        let date = Date(timeIntervalSince1970: unixtime)
        return formatter.string(from: date)
    }

    private let formatter = DateFormatter()
    @State private var safariOpenUrl: URL? = nil

    init(post: Post, onPressItem: ((_ post: Post) -> Void)? = nil) {
        self.post = post
        self.onPressItem = onPressItem
        formatter.dateFormat = "yyyy/MM/dd"
        formatter.locale = Locale(identifier: "ja_JP")
    }

    var body: some View {
        Button(action: {
            if let onPressItem {
                onPressItem(post)
                return
            }
            safariOpenUrl = URL(string: content.url ?? "")
        }) {
            VStack(alignment: .leading) {
                HStack {
                    if let imageUrl = poster?.url {
                        AsyncImage(url: URL(string: imageUrl)) { image in
                            image.resizable()
                                .scaledToFit()
                        } placeholder: {
                            ProgressView()
                        }
                        .frame(width: 112, height: 88)
                    } else {
                        Image(.noimage)
                            .resizable()
                            .frame(width: 112, height: 88)
                    }
                    VStack(alignment: .leading) {
                        Text(content.title ?? "")
                            .font(.system(size: 20))
                            .bold()
                            .foregroundColor(Color(.text))
                            .lineLimit(nil)
                            .padding(.bottom, 16)
                        Text(hostname)
                            .font(.body)
                            .foregroundColor(Color(.subText))
                    }
                    .frame(alignment: .top)
                }
                HStack {
                    ForEach(post.tags, id: \.self) { tag in
                        Text("#\(tag)")
                            .padding(.vertical, 4)
                            .padding(.horizontal, 8)
                            .foregroundColor(Color(.text))
                            .background(Color(.lightGray))
                            .cornerRadius(5)
                    }
                }
                Divider()
                    .padding(.vertical, 16)
                Text(formattedDate)
                    .foregroundColor(Color(.subText))
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.vertical, 32)
            .background(Color(.blurGray))
            .cornerRadius(15)
            .background(Color.clear)
        }
        .openSafariView(url: safariOpenUrl) {
            safariOpenUrl = nil
        }
    }
}

#Preview {
    let post = Post(
        idString: "111",
        timestamp: 1480134638,
        tags: ["blog", "contribution"],
        content: [
            PostContent(
                type: "link",
                description: "test description",
                title: "test title abc def ghi jkl mno pqr stu vwx yz - 123 456 789 0",
                url: "https://asmz.beer",
                poster: [
                    PostContentPoster(url: "https://static.tumblr.com/nibuxm5/Q0yo2heyn/beer.jpg")
                ]
            )
        ]
    )

    return PostItem(post: post)
}
